Fix alpha API warnings for patch version differences

akhilsingh-git opened the pull request:
This prevents spurious warnings when emulation version is set to major.minor (e.g., 1.34) and binary version includes patch (e.g., 1.34.1).

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug 


#### What this PR does / why we need it:
Fixes issue #134023 where alpha API warnings were being logged when binary version (1.34.1) and emulation version (1.34) differed only in patch version.

The issue was in api_enablement.go where the version comparison was using EqualTo() which compares all version components including patch versions. The fix changes the comparison to only check major.minor versions using version.MajorMinor().

#### Special notes for your reviewer:
Changes:
- Modified version comparison logic in ApplyTo() method to only compare major.minor versions, not patch versions
- Added comprehensive test cases to verify the fix works correctly
- Tests confirm that warnings are still logged for different major/minor versions but not for different patch versions

akhilsingh-git committed:
Fix alpha API warnings for patch version differences

Fixes issue #134023 where alpha API warnings were being logged
when binary version (1.34.1) and emulation version (1.34) differed
only in patch version.

The issue was in api_enablement.go where the version comparison
was using EqualTo() which compares all version components including
patch versions. The fix changes the comparison to only check
major.minor versions using version.MajorMinor().

Changes:
- Modified version comparison logic in ApplyTo() method to only
  compare major.minor versions, not patch versions
- Added comprehensive test cases to verify the fix works correctly
- Tests confirm that warnings are still logged for different
  major/minor versions but not for different patch versions

This prevents spurious warnings when emulation version is set to
major.minor (e.g., 1.34) and binary version includes patch (e.g., 1.34.1).

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/api_enablement.go:
```diff
@@ -106,7 +107,7 @@ func (s *APIEnablementOptions) ApplyTo(c *server.Config, defaultResourceConfig *
 	c.MergedResourceConfig = mergedResourceConfig
 
-	if binVersion, emulatedVersion := c.EffectiveVersion.BinaryVersion(), c.EffectiveVersion.EmulationVersion(); !binVersion.EqualTo(emulatedVersion) {
+	if binVersion, emulatedVersion := c.EffectiveVersion.BinaryVersion(), c.EffectiveVersion.EmulationVersion(); !version.MajorMinor(binVersion.Major(), binVersion.Minor()).EqualTo(version.MajorMinor(emulatedVersion.Major(), emulatedVersion.Minor())) {
```
```suggestion
	if binVersion, emulatedVersion := c.EffectiveVersion.BinaryVersion(), c.EffectiveVersion.EmulationVersion(); binVersion.Major() != emulatedVersion.Major() || binVersion.Minor() != emulatedVersion.Minor() {
```

akhilsingh-git replied:
i have updated the same

akhilsingh-git committed:
Address PR feedback: simplify version comparison and improve test coverage

- Apply liggitt's suggestion to use direct major/minor comparison instead of MajorMinor objects
- Remove unused version import
- Address AvineshTripathi's feedback by adding comprehensive test coverage with log verification
- Add proper test cases that verify warning behavior and error conditions
- Include failing test cases as requested

This ensures the fix for spurious alpha API warnings is robust and well-tested.

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>

akhilsingh-git commented on staging/src/k8s.io/apiserver/pkg/server/options/api_enablement_test.go:
```diff
@@ -77,3 +83,165 @@ func TestAPIEnablementOptionsValidate(t *testing.T) {
+	return nil
+}
+
+func TestAPIEnablementOptionsApplyToVersionComparison(t *testing.T) {
```
done

akhilsingh-git committed:
Fix golangci-lint issue: simplify else-if statement

Replace 'else {if cond {}}' with 'else if cond {}' as suggested by gocritic linter.

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>

akhilsingh-git committed:
Fix gofmt formatting issues

Remove trailing whitespace and align struct fields consistently.

Signed-off-by: Akhil Singh <singhakhil69@gmail.com>

liggitt commented on staging/src/k8s.io/apiserver/pkg/server/options/api_enablement.go:
```diff
@@ -106,7 +106,7 @@ func (s *APIEnablementOptions) ApplyTo(c *server.Config, defaultResourceConfig *
-	if binVersion, emulatedVersion := c.EffectiveVersion.BinaryVersion(), c.EffectiveVersion.EmulationVersion(); !binVersion.EqualTo(emulatedVersion) {
+	if binVersion, emulatedVersion := c.EffectiveVersion.BinaryVersion(), c.EffectiveVersion.EmulationVersion(); binVersion.Major() != emulatedVersion.Major() || binVersion.Minor() != emulatedVersion.Minor() {
 		for _, version := range registry.PrioritizedVersionsAllGroups() {
 			if strings.Contains(version.Version, "alpha") {
```
@michaelasp is everything PrioritizedVersionsAllGroups() returns even necessarily enabled? Should this be checking enablement in the MergedResourceConfig before warning, something like this:

```go
			entireVersionEnabled := c.MergedResourceConfig.ExplicitGroupVersionConfigs[version]
			individualResourceEnabled := false
			for resource, enabled := range c.MergedResourceConfig.ExplicitResourceConfigs {
				if enabled && resource.Group == version.Group && resource.Version == version.Version {
					individualResourceEnabled = true
					break
				}
			}
			if (entireVersionEnabled || individualResourceEnabled) && strings.Contains(version.Version, "alpha") {
```

michaelasp replied:
If they're registered then the APIs are enabled no? If not then yeah we should check whether this is enabled here too. I was under the assumption that if the APIs themselves are registered then it's enabled. Otherwise, we should add that here, cc @Jefftree as well to chime in.

liggitt replied:
> I was under the assumption that if the APIs themselves are registered then it's enabled

I don't think so.

https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/instance.go#L504-L506 registers alphas and betas explicitly but as disabled

michaelasp replied:
Gotcha, yeah in that case we should also add an enablement check here. Good catch! We can check whether the api is enabled using the `ExplicitResourceConfigs`

siyuanfoundation replied:
That is a bug. Good catch!

michaelasp replied:
I can fix this in a separate PR since it's not exactly related. @akhilsingh-git let me know, otherwise I'll send out a fix for that by EOW.

michaelasp replied:
Thanks for the fix!

siyuanfoundation commented on staging/src/k8s.io/apiserver/pkg/server/options/api_enablement_test.go:
```diff
@@ -77,3 +85,271 @@ func TestAPIEnablementOptionsValidate(t *testing.T) {
+	return f.versions
+}
+
+type fakeEffectiveVersion struct {
```
Why is do you need a `fakeEffectiveVersion` instead of just using `basecompatibility.EffectiveVersion`?

akhilsingh-git committed:
Fix alpha API warnings for patch version differences

Fixes issue #134023 where alpha API warnings were being logged
when binary version (1.34.1) and emulation version (1.34) differed
only in patch version.

The issue was in api_enablement.go where the version comparison
was using EqualTo() which compares all version components including
patch versions. The fix changes the comparison to only check
major.minor versions using version.MajorMinor().

Changes:
- Modified version comparison logic in ApplyTo() method to only
  compare major.minor versions, not patch versions
- Added enablement check for alpha APIs before warning using
  MergedResourceConfig to ensure warnings are only shown for
  actually enabled alpha APIs
- Comprehensive test cases already exist to verify the fix works correctly

This prevents spurious warnings when emulation version is set to
major.minor (e.g., 1.34) and binary version includes patch (e.g., 1.34.1).

michaelasp commented:
/retest

michaelasp commented:
/lgtm

This is a usability bugfix, lets try and get this in. @akhilsingh-git is there anything else left? Looks good on my end.

pohly commented:
This can be closed because it was merged through https://github.com/kubernetes/kubernetes/pull/135327, right?
